File: game.zig
const std = @import("std");

const builtin = std.builtin;

const allocator = std.heap.page_allocator; //it maps directly to WebAssembly memory.

//based on https://github.com/daneelsan/zig-wasm-logger
const Imports = struct {
    extern "env" fn _throwError(pointer: [*]const u8, length: u32) noreturn;
    pub fn throwError(message: []const u8) noreturn {
        _throwError(message.ptr, message.len);
    }
    extern fn jsConsoleLogWrite(ptr: [*]const u8, len: usize) void;
    extern fn jsConsoleLogFlush() void;
    extern fn jsAskForString(pointer: [*]const u8) [*:0]u8; //slice of u8
};

pub const Console = struct {
    pub const Logger = struct {
        pub const Error = error{};
        pub const Writer = std.io.Writer(void, Error, write);

        fn write(_: void, bytes: []const u8) Error!usize {
            Imports.jsConsoleLogWrite(bytes.ptr, bytes.len);
            return bytes.len;
        }
    };

    const logger = Logger.Writer{ .context = {} };
    pub fn log(comptime format: []const u8, args: anytype) void {
        logger.print(format, args) catch return;
        Imports.jsConsoleLogFlush();
    }
};

export fn allocString(length: u32) [*]const u8 {
    const slice = allocator.allocSentinel(u8, length, 0) catch
        @panic("failed to allocate memory");
    //note: 0 is a valid WASM pointer
    return slice.ptr;
}

export fn alloc(length: u32) [*]const u8 {
    const slice = allocator.alloc(u8, length) catch
        @panic("failed to allocate memory");
    return slice.ptr;
}

export fn free(pointer: [*:0]u8) void {
    allocator.free(std.mem.span(pointer));
}

fn to_slice(pointer: [*]u8, len: usize) []u8 {
    return pointer[0..len];
}

// Calls to @panic are sent here.
// See https://ziglang.org/documentation/master/#panic
pub fn panic(message: []const u8, _: ?*builtin.StackTrace, _: ?usize) noreturn {
    Imports.throwError(message);
}

export fn main() void {
    //std.debug.print("Hello, {s}!\n", .{"World"});
    Console.log("Hello, {s}!\n", .{"World"});
}

//no string arguments in WASM land :(
//instead, accepts a string pointer (linear memory offset)
export fn update(message_pointer: [*:0]const u8) void {
    //only accepts slices so we need to transform pointer into a slice
    defer allocator.free(std.mem.sliceTo(message_pointer, 0)); //we need to deallocate on Zig side since we allocated on JS side above
    const name: []const u8 = std.mem.span(message_pointer); //get Zig-style slice
    Console.log("Hi, {s}\n", .{name});
}

export fn sumArray(ptr: [*]u8) void {
    var len: usize = ptr[0]; //first entry encodes length (think BSON)
    const array: []const u8 = to_slice(ptr, len);
    Console.log("sum: {any}\n", .{_sumArray(array)});
}

fn _sumArray(array: []const u8) u8 {
    var total: u8 = 0;

    for (array) |byte| {
        total += byte;
    }

    return total;
}

export fn parse(message_pointer: [*:0]const u8) void {
    //only accepts slices so we need to transform pointer into a slice
    defer allocator.free(std.mem.sliceTo(message_pointer, 0)); //we need to deallocate on Zig side since we allocated on JS side above
    const script_payload: []const u8 = std.mem.span(message_pointer); //get Zig-style slice
    Console.log("Script: {s}\n", .{script_payload});
}

File: game.js
// game.js - load the WASM, display the game, update the engine

let console_log_buffer = "";
//let memory = null; //hack for decodeString to work
//let allocUint8 = null;

//based on https://github.com/daneelsan/zig-wasm-logger/blob/master/script.js
let wasm = {
    instance: undefined,

    init: function (obj) {
        this.instance = obj.instance;
    },
    decodeString: function(pointer, length){
        const memory = this.instance.exports.memory;
        const slice = new Uint8Array(
            memory.buffer,
            pointer,
            length
          )
          return new TextDecoder().decode(slice)
    },
    encodeString: function(string){
        const bytes = new TextEncoder().encode(string)
        const len = bytes.byteLength
        //this can potentially invalidate the memory buffer
        const ptr = this.instance.exports.allocString(len) // ask Zig to allocate memory
        // console.log("Pointer hex: ", "0x"+ptr.toString(16)); //since it's a pointer, no negative values are possible
        //..so we have to always do new/recreate here
        const view = this.getMemory()
        view.subarray(ptr, ptr + len).set(bytes)

        return ptr
    },
    getMemory: function () {
        const memory = this.instance.exports.memory
        if (
            this.wasmMemory === undefined ||
            this.wasmMemory !== memory.buffer
        ) {
            this.wasmMemory = new Uint8Array(memory.buffer)
        }
        return this.wasmMemory
    },
    encodeBytes: function(){
        var input = new Uint8Array([1, 2, 3, 4, 5]);
        const length = input.byteLength;
        var bytes = new Uint8Array(length+1);
        // make the first entry encode length
        bytes[0] = length;
        bytes.set(input, 1);
        console.log(bytes);
        const ptr = this.instance.exports.alloc(length+1); //ask Zig to allocate

        const view = this.getMemory()
        view.subarray(ptr, ptr + length+1).set(bytes) //begin is inclusive, end is exclusive
        //this.u8a(ptr + this.LENGTH_WIDTH, length).set(bytes);
        console.log("Pointer hex: ", "0x"+ptr.toString(16)); //since it's a pointer, no negative values are possible
        return ptr;
    }
}

//any JS functions we want to expose to WASM go here
var envObject = { env: {
    //initialize memory here
    memoryBase: 0,
    tableBase: 0,
    memory: new WebAssembly.Memory({ initial: 512 }),
    _throwError(pointer, length) {
        const message = wasm.decodeString(pointer, length)
        throw new Error(message)
    },
    jsConsoleLogWrite: function (ptr, len) {
        console_log_buffer += wasm.decodeString(ptr, len);
    },
    jsConsoleLogFlush: function () {
        console.log(console_log_buffer);
        console_log_buffer = "";
    },
    jsAskForString: function(ptr, len) {
        return wasm.decodeString(ptr, len);
    }
} };

//those will hold App and its state
//App is WASM exports which is immutable, hence separate AppState
var App = {
};

var AppState = {
    'loaded': false,
    'running': false,
};

async function bootstrap() {
    

    wasm.init(await WebAssembly.instantiateStreaming(fetch("game.wasm"), envObject));
    
    //.then(result => {
    if (wasm != null){ 
    console.log("Loaded the WASM!");
    //App = wasm.instance.exports;
    console.log(wasm);
    AppState.loaded = true;
    AppState.running = true;
    //App.init();
    wasm.instance.exports.main(); // begin

    document.getElementById("1").onclick = function(){
        if (wasm != null) {
            var str = wasm.encodeString("test")
            console.log(str)
            wasm.instance.exports.update(str)
            //test
            var arr_ptr = wasm.encodeBytes();
            wasm.instance.exports.sumArray(arr_ptr)
        }
    }
    
    document.getElementById("2").onclick = function(){
        if (wasm != null){
            var str = wasm.encodeString("other")
            console.log(str)
            wasm.instance.exports.update(str)
        }
    }

    } //);
}

window.document.body.onload = function() {
   bootstrap()
};

